### ISSUE TITLE
Option -r / --regions off-by-1: returns indels one past the end of the region

### ISSUE THREAD

**Issue by jmarshall** [samtools/bcftools#1420]
Consider [events.vcf](https://github.com/samtools/bcftools/files/6005716/events.vcf.txt), which ends with the following VCF records:

```
… …
chr1	199	s3	A	T	.	PASS	.
chr1	199	d5f	AT	A	.	PASS	.
chr1	199	i4f	A	ATTTT	.	PASS	.
chr1	200	s4f	A	T	.	PASS	.
chr1	200	outd2	AT	A	.	AFTER	.
chr1	200	outi4	A	AT	.	AFTER	.
chr1	201	outs2	A	T	.	AFTER	.
chr1	201	outd3	AT	A	.	AFTER	.
chr1	201	outi5	A	ATTTT	.	AFTER	.
```

If we query this for `chr1:100-200`, we would expect to receive the `PASS` records but not the `AFTER` records. In particular for the records with POS=200, we would expect get `s4f` back as the base at position 200 has been changed, but IMNSHO not `outd2` or `outi4` as they only affect bases to the right of position 200 so are outwith the specified region.

However bcftools's index-based `-r`/`--regions`/`-R`/`--regions-file` query (both as released and on **develop**) returns all three of these records:

```
$ bgzip events.vcf && bcftools index events.vcf.gz
$ bcftools view -r chr1:100-200 events.vcf.gz
##fileformat=VCF4.3
… …
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO
… …
chr1	199	s3	A	T	.	PASS	.
chr1	199	d5f	AT	A	.	PASS	.
chr1	199	i4f	A	ATTTT	.	PASS	.
chr1	200	s4f	A	T	.	PASS	.
chr1	200	outd2	AT	A	.	AFTER	.
chr1	200	outi4	A	AT	.	AFTER	.

```

**Comment by jkbonfield** [comment]
It depends on the question being asked - is it to return records whose POS field is within this range or records whose base changes are within this range?

I'm curious to know how other tools handle this scenario.  Tabix would clearly return those two records, and making it do anything else would be almost impossible as it's a general purpose query that doesn't understand things like VCF indel syntax.  I can't find anything in Picard that can subset VCFs, and GATK SelectVariants just tells me "no suitable codecs found".  I've never quite got to grips with the appropriate hoops to jump there. :(

**Comment by jmarshall** [comment]
That is true. There is a purely computational question about overlapping with the coordinates subtended by the VCF record, and there is the scientifically meaningful question about finding events that overlap with the specified region. Personally I expect `bcftools view` to facilitate the scientific question first…

**Comment by jkbonfield** [comment]
Yes, but it's not helpful to users if different tools use different methods of assessing overlaps, hence why I say we need to look at other programs.  It helps no one to have yet more variation across bioinformatics tools.

Alas my arcane knowledge is insufficient.  If you know the magic runes required to get GATK to do anything then give it a whirl.  I got nowhere.  I also couldn't get Picard to swallow it, regardless what version I put in the header.  It simply kept saying "VCF4."whatever isn't a valid version, but it wouldn't tell me what versions it *does* support.

**Comment by pd3** [comment]
One note, records such as
```
1	200	.	AT	A
```
where the true variation starts after the end of the region (e.g. `1:100-200`) should be reported as well. Otherwise, if we insisted on the program reporting only real variation, then records with empty ALT column such as 
```
1	200	.	A	.
```
should be never printed, which is obviously not what we want.

**Comment by jmarshall** [comment]
The raison d'être of this bug report is that `1  200  .  AT  A` does not touch the base at 200 therefore should NOT be reported by a query for `1:100-200`.

The definition of what bases `1 200 . A .` should be considered to touch is a separate question. Normally in VCF both REF and ALT are required to be non-empty; `ALT=.` is a special case. The spec says `ALT=.` indicates “no variant [at that position, presumably]” so I think programs should consider it to touch the single base in the range 200–200. Hence it would be reported by a query for `1:100-200`.

(This bug report is that `-r` returns slightly too many hits due to the “touched bases” for indels being slightly too generous. It is a minor bug compared to #1421, in which `-t` returns too few hits — compared to the expectations of users who didn't read the documentation extremely carefully.)

**Comment by pd3** [comment]
Handling indels is difficult and it is not always clear what is the right thing to do. For example, if the deleted T is part of a homopolymer run, we cannot distinguish which of the T's was actually deleted. Some would argue, as you do, that "it should NOT be reported", others say that it "MUST" be reported. There is no single universally correct approach, the only solution is to make the behavior optional. I have a half-finished code that I put aside a while ago, I may revisit it.

**Comment by jkbonfield** [comment]
Given https://github.com/samtools/htslib/pull/1327 is now merged and the corresponding bcftools commits are in (https://github.com/samtools/bcftools/commit/0d04159437d12de6f9c672c18347c60591e3f458) we believe this to be fixed.

**Comment by jmarshall** [comment]
Has the bcftools commit been reviewed? Did any review have an opportunity to comment on the obscure nature of the options added to control this?

```
--regions-overlap 0|1|2       Include if POS in the region (0), record overlaps (1),
                              variant overlaps (2) [1]
```

**Comment by pd3** [comment]
No, it has not. If you don't like the obscure nature of bcftools, please don't use it.
